### PR TITLE
Add workflow code coverage report and upload

### PR DESCRIPTION
**Pull request by jtmoon79**
Add github workflow to run grcov code coverage over `cargo test` and upload report to codecov.io

Here is the report for this commit https://app.codecov.io/github/chronotope/chrono/commit/b46ac6c090f1832e1c73903318098067d471737b/tree
Comparison data is missing because there have been no prior code coverage runs. This URL can be found in workflow _grcov_, step _upload codecov_, in the last few logged lines. e.g.
```text
[2023-07-09T06:36:32.407Z] ['info'] {"status":"success","resultURL":"https://app.codecov.io/github/chronotope/chrono/commit/0e251e6b2764ec2f736300f925b5e22b09c58bd4"}
```
The `README.md` also has a badge+link to the latest codecov.io code coverage report.

<br/>

For some reason, the newly added _grcov_ workflow does not show up in some _Actions_ views. I can find it by clicking _Actions_ (at the top), then click _grcov_ (left-side), then the runs of `grcov.yml` are listed.

### CODE REVIEW

**Comment by pitdicker** [comment]
How does this compare to `cargo-llvm-cov`? https://github.com/taiki-e/cargo-llvm-cov#continuous-integration

**Comment by jtmoon79** [comment]
I added `cargo-llvm-cov` job to my own project.

### `cargo-llvm-cov` and coveralls.io

Here's my code for a `cargo-llvm-cov` job https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/9948f0a9035b0883644f0a37d63d16a77158be5b/.github/workflows/rust.yml#L543-L583

Run time is 8m10s. https://github.com/jtmoon79/super-speedy-syslog-searcher/actions/runs/5516276662

I tried out coveralls.io as a renderer service. Here is an example of the output https://coveralls.io/builds/61257604

### `grcov` and codecov.io

Here's my code for `grcov` job https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/9948f0a9035b0883644f0a37d63d16a77158be5b/.github/workflows/rust.yml#L445-L540 (similar to this PR)

Run time is 23m59s. https://github.com/jtmoon79/super-speedy-syslog-searcher/actions/runs/5516276662

Here is the codecov.io rendering of that output https://app.codecov.io/gh/jtmoon79/super-speedy-syslog-searcher/commit/93a58cc0254e8f4965fe7e3d5cd702489a237ee0/tree

I don't currently know why the coverage differences are different between `cargo-llvm-cov` and `grcov`.

## Which one?

Which do you prefer @pitdicker @djc ?

`cargo-llvm-cov` is simpler to setup and run, and is faster than `grcov`.

I found the rendering at coveralls.io better looking and a bit easier to navigate. Both are decent, and AFAIK, permanently free for open-source projects.

**Comment by pitdicker** [comment]
Great you looked at both! The highlighted lines in coveralls.io make slightly more sense to me.

I suppose it is the same as with benchmarking: you can spend a lot of time trying to get more accurate. But really the useful part is doing a measurement and having something to compare against.

The CI run for with grcov for this PR took ca. 5 minutes (https://github.com/chronotope/chrono/actions/runs/5498456361/jobs/10019932092?pr=1178), most of which is spend in `cargo install grcov`. It seems the runtime does not have to be a concern for us? Although it is nice `cargo-llvm-cov` is the faster one.

**Comment by pitdicker** [comment]
Is one of them easier to replicate 'at home'? And do we have a place to document how to do so?

**Comment by jtmoon79** [comment]
> I suppose it is the same as with benchmarking: you can spend a lot of time trying to get more accurate. But really the useful part is doing a measurement and having something to compare against.

Yep.

> Is one of them easier to replicate 'at home'? And do we have a place to document how to do so?

I think you're asking "which is easier to use?" ?  If so, then `cargo-llvm-cov` is definitely easier, as you can see from the amount of setup in each github Action job.

**Comment by djc** [comment]
I prefer cargo-llvm-cov over grcov. It is my impression that coveralls is no longer being actively developed? CodeCov hasn't  exactly been without issues but it was recently acquired by Sentry who I have a decent amount of trust in.

**Comment by pitdicker** [comment]
I have no idea about either, but https://coveralls.io/announcements seems pretty active.

**Comment by jtmoon79** [comment]
I changed to use `llvm-cov` and codecov.io

File: .github/workflows/codecov.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: ${{ env.MSRV }}
+          override: true
+      - name: install cargo-llvm-cov
+        uses: actions-rs/cargo@v1
+        with:
+          command: install
+          args: cargo-llvm-cov --version ${{ env.LLVM_COV_VERSION }}
```
Comment (pitdicker) [inline .github/workflows/codecov.yml:30]: Why are we not using the pre-build binaries as recommended for GitHub Actions? https://github.com/taiki-e/cargo-llvm-cov#on-github-actions

File: .github/workflows/codecov.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    branches: [main, 0.4.x]
+  pull_request:
+jobs:
+  # Run code coverage using cargo-llvm-cov then upload to codecov.io
+  job_code_coverage:
+    name: llvm-cov
+    runs-on: ubuntu-latest
+    env:
+      # `MSRV` should match project `cargo-llvm-cov` value for `rust-version`
+      MSRV: 1.64
```
Comment (pitdicker) [inline .github/workflows/codecov.yml:13]: The oldest version supported with pre-build binaries is 1.60.
But is it not best to just use the latest rust version?

Comment (jtmoon79) [reply .github/workflows/codecov.yml]: Removed the `MSRV` setting

File: .github/workflows/codecov.yml (hunk 1 above)
Comment (jtmoon79) [reply .github/workflows/codecov.yml]: Added `- uses: taiki-e/install-action@cargo-llvm-cov`

**Review by pitdicker — APPROVED** [review]

File: .github/workflows/codecov.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  pull_request:
+jobs:
+  # Run code coverage using cargo-llvm-cov then upload to codecov.io
+  job_code_coverage:
+    name: llvm-cov
+    runs-on: ubuntu-latest
+    env:
+      COVERAGE_FILE: ./coverage.lcov
+    steps:
+      - uses: taiki-e/install-action@cargo-llvm-cov
```
Comment (djc) [inline .github/workflows/codecov.yml:14]: I'd like this to stick as closely as possible to the GH Actions example in the cargo-llvm-cov README, or have comments that explain any required differences.

**Comment by jtmoon79** [comment]
https://app.codecov.io/github/chronotope/chrono/commit/6db2eb94a29243286b783c614ac22fd216e058ba

**Comment by djc** [comment]
Thanks!

**Pull request merged**